File: python_open/perspectiveTransfrom.py
import cv2
import numpy as np

folder = "/home/test/Desktop/opencv/opencv/cppOpen/build/data/"
src = cv2.imread(folder + "img9.bmp", cv2.IMREAD_COLOR)

if src is None:
    print("Error: Could not load image!")
    exit(-1)

# 전역 변수
src_pts = []
dst_pts = np.float32([[0, 0], [199, 0], [199, 299], [0, 299]])
cnt = 0  # 점의 개수 카운트

def on_mouse(event, x, y, flags, param):
    global cnt
    if event == cv2.EVENT_LBUTTONDOWN:
        if cnt < 4:  # 최대 4개의 점을 선택
            src_pts.append((x, y))
            cnt += 1
            print(f"Point added: ({x}, {y})")
            cv2.circle(src, (x, y), 10, (255, 0, 0), -1)  # 선택한 점을 원으로 표시
            cv2.imshow("src", src)  # 업데이트된 이미지 보여주기

            if cnt == 4:
                # 원근 변환 행렬 계산
                M = cv2.getPerspectiveTransform(np.float32(src_pts), dst_pts)
                dst = cv2.warpPerspective(src, M, (200, 300))  # 원근 변환 적용
                cv2.imshow("dst", dst)
                cnt = 0  # 점 선택 카운터 초기화

# 윈도우 및 콜백 함수 설정
cv2.namedWindow("src")
cv2.setMouseCallback("src", on_mouse)

cv2.imshow("src", src)
cv2.waitKey(0)
cv2.destroyAllWindows()

File: cppOpen/perspectiveTransfrom.cpp
#include "opencv2/opencv.hpp"
#include <iostream>

using namespace cv;
using namespace std;

String folder = "/home/test/Desktop/opencv/opencv/cppOpen/build/data/";
void onMouse(int event, int x, int y, int flags, void *data);

Mat src;
Point2f srcPts[4], dstPts[4];
int cnt = 0; // 전역으로 cnt를 관리

int main()
{
    src = imread(folder + "img9.bmp", IMREAD_COLOR);
    if (src.empty()) {
        cout << "이미지를 열 수 없거나 찾을 수 없습니다!" << endl;
        return -1;
    }

    namedWindow("src");
    setMouseCallback("src", onMouse);
    imshow("src", src);
    waitKey();
    return 0;
}

void onMouse(int event, int x, int y, int flags, void *data)
{
    switch (event)
    {
    case EVENT_LBUTTONDOWN:
        if (cnt < 4) { // cnt가 4보다 작은 경우에만
            srcPts[cnt++] = Point2f(x, y);
            cout << "Point added: (" << x << ", " << y << ")" << endl;
            circle(src, Point(x, y), 10, Scalar(255, 0, 0), -1);
            imshow("src", src); // 원본 이미지를 업데이트

            if (cnt == 4)
            {
                dstPts[0] = Point2f(0, 0);
                dstPts[1] = Point2f(199, 0);
                dstPts[2] = Point2f(199, 299);
                dstPts[3] = Point2f(0, 299);

                Mat M = getPerspectiveTransform(srcPts, dstPts);
                Mat dst;
                warpPerspective(src, dst, M, Size(200, 300));
                imshow("dst", dst);
                cnt = 0; // 점 추가 카운터 초기화
            }
        } else {
            cout << "4개의 점이 이미 선택되었습니다." << endl;
        }
        break;
    }
}
